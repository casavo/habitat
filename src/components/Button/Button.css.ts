/* eslint-disable sort-keys-fix/sort-keys-fix */
import { recipe } from "@vanilla-extract/recipes";
import { sprinkles } from "../../utils/sprinkles.css";
import { vars } from "../../utils/theme.css";
import { keyframes, style } from "@vanilla-extract/css";

const rotate = keyframes({
  "0%": { transform: "rotate(0deg)" },
  "100%": { transform: "rotate(360deg)" },
});

// or interpolate as a shorthand:
export const infiniteRotation = style(
  {
    animation: `${rotate} 0.6s ease-out infinite`,
  },
  "rotation"
);

export const ButtonContainer = recipe(
  {
    base: [
      sprinkles({
        gap: 100,
        pt: 150,
        pb: 150,
        pr: 200,
        pl: 200,
        d: "inline-flex",
      }),
      {
        alignItems: "center",
        alignSelf: "stretch",
        borderRadius: 4,
        justifyContent: "center",
        cursor: "pointer",
        ":disabled": {
          cursor: "not-allowed",
        },
        ":focus-visible": {
          outline: "none",
        },
      },
    ],
    variants: {
      theme: {
        dark: {},
        light: {},
      },
      variant: {
        primary: {},
        secondary: {},
        tertiary: {},
      },
      size: {
        small: {
          height: 40,
        },
        medium: {
          height: 48,
        },
        large: {
          height: 54,
        },
      },
      iconLayout: {
        sx: {
          flexDirection: "row",
        },
        dx: {
          flexDirection: "row-reverse",
        },
      },
    },
    compoundVariants: [
      {
        variants: {
          theme: "dark",
          variant: "primary",
        },
        style: {
          backgroundColor: vars.colors.neutral[100],
          color: vars.colors.neutral[0],
          fill: vars.colors.neutral[0],
          border: "none",
          ":hover": {
            backgroundColor: vars.colors.neutral[60],
          },
          ":active": {
            backgroundColor: vars.colors.neutral[80],
          },
          ":disabled": {
            backgroundColor: vars.colors.neutral[10],
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
          },
        },
      },
      {
        variants: {
          theme: "light",
          variant: "primary",
        },
        style: {
          backgroundColor: vars.colors.neutral[0],
          color: vars.colors.neutral[100],
          fill: vars.colors.neutral[100],
          border: "none",
          ":hover": {
            backgroundColor: vars.colors.neutral[10],
          },
          ":active": {
            backgroundColor: vars.colors.neutral[20],
          },
          ":disabled": {
            backgroundColor: vars.colors.neutral[10],
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
          },
        },
      },
      {
        variants: {
          theme: "dark",
          variant: "secondary",
        },
        style: {
          backgroundColor: vars.colors.neutral[0],
          color: vars.colors.neutral[100],
          fill: vars.colors.neutral[100],
          border: `1px solid ${vars.colors.neutral[100]}`,
          ":hover": {
            backgroundColor: vars.colors.neutral[100],
            color: vars.colors.neutral[0],
            fill: vars.colors.neutral[0],
          },
          ":active": {
            color: vars.colors.neutral[0],
            fill: vars.colors.neutral[0],
            backgroundColor: vars.colors.neutral[80],
          },
          ":disabled": {
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
            backgroundColor: vars.colors.neutral[0],
            border: `1px solid ${vars.colors.neutral[60]}`,
          },
        },
      },
      {
        variants: {
          theme: "light",
          variant: "secondary",
        },
        style: {
          backgroundColor: vars.colors.neutral[100],
          color: vars.colors.neutral[0],
          fill: vars.colors.neutral[0],
          border: `1px solid ${vars.colors.neutral[0]}`,
          ":hover": {
            backgroundColor: vars.colors.neutral[0],
            color: vars.colors.neutral[100],
            fill: vars.colors.neutral[100],
          },
          ":active": {
            backgroundColor: vars.colors.neutral[10],
            color: vars.colors.neutral[100],
            fill: vars.colors.neutral[100],
          },
          ":disabled": {
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
            backgroundColor: vars.colors.neutral[100],
            border: `1px solid ${vars.colors.neutral[60]}`,
          },
        },
      },
      {
        variants: {
          theme: "dark",
          variant: "tertiary",
        },
        style: {
          backgroundColor: "transparent",
          color: vars.colors.neutral[100],
          fill: vars.colors.neutral[100],
          ":hover": {
            backgroundColor: vars.colors.neutral[100],
            color: vars.colors.neutral[0],
            fill: vars.colors.neutral[0],
          },
          ":active": {
            backgroundColor: vars.colors.neutral[80],
            color: vars.colors.neutral[0],
            fill: vars.colors.neutral[0],
          },
          ":disabled": {
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
            backgroundColor: "transparent",
          },
        },
      },
      {
        variants: {
          theme: "light",
          variant: "tertiary",
        },
        style: {
          backgroundColor: "transparent",
          color: vars.colors.neutral[0],
          fill: vars.colors.neutral[0],
          ":hover": {
            backgroundColor: vars.colors.neutral[0],
            color: vars.colors.neutral[100],
            fill: vars.colors.neutral[100],
          },
          ":active": {
            backgroundColor: vars.colors.neutral[10],
            color: vars.colors.neutral[100],
            fill: vars.colors.neutral[100],
          },
          ":disabled": {
            color: vars.colors.neutral[60],
            fill: vars.colors.neutral[60],
            backgroundColor: "transparent",
          },
        },
      },
    ],
  },
  ""
);
