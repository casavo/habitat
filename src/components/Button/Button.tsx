import { useButton } from "react-aria";
import { ButtonHTMLAttributes, useRef } from "react";

type Props = {
  children: React.ReactNode;
  "data-testid"?: string;
  type?: ButtonHTMLAttributes<object>["type"];
};

export const Button = (props: Props) => {
  const ref = useRef<HTMLButtonElement>(null);
  const { buttonProps } = useButton(props, ref);
  const { children } = props;

  return (
    <button {...buttonProps} ref={ref}>
      {children}
    </button>
  );
};
