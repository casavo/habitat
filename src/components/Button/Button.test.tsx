import { render } from "@testing-library/react";
import { Button } from "./Button.tsx";
import { expect, describe } from "vitest";

describe("Button component", () => {
  it("renders", () => {
    const { container } = render(<Button>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with a theme", () => {
    const { container } = render(<Button theme="dark">Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with a size", () => {
    const { container } = render(<Button size="large">Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with a disabled state", () => {
    const { container } = render(<Button disabled>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });

  it("renders with a loading state", () => {
    const { container } = render(<Button loading>Click me</Button>);
    expect(container.firstChild).toMatchSnapshot();
  });
});
