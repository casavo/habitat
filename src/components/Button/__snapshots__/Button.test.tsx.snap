// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Button component > renders 1`] = `
<button
  class="Button_ButtonContainer__1kgv65v2 Button_ButtonContainer_iconLayout_sx__1kgv65vb Button_ButtonContainer_size_medium__1kgv65v9 Button_ButtonContainer_theme_dark__1kgv65v3 Button_ButtonContainer_variant_primary__1kgv65v5 Button_ButtonContainer_compound_0__1kgv65vd"
  data-rac=""
  type="button"
>
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne Typography_BodyStyle_strong_true__3igy4ng"
  >
    Click me
  </p>
</button>
`;

exports[`Button component > renders with a disabled state 1`] = `
<button
  class="Button_ButtonContainer__1kgv65v2 Button_ButtonContainer_iconLayout_sx__1kgv65vb Button_ButtonContainer_size_medium__1kgv65v9 Button_ButtonContainer_theme_dark__1kgv65v3 Button_ButtonContainer_variant_primary__1kgv65v5 Button_ButtonContainer_compound_0__1kgv65vd"
  data-disabled="true"
  data-rac=""
  disabled=""
  type="button"
>
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne Typography_BodyStyle_strong_true__3igy4ng"
  >
    Click me
  </p>
</button>
`;

exports[`Button component > renders with a loading state 1`] = `
<button
  class="Button_ButtonContainer__1kgv65v2 Button_ButtonContainer_iconLayout_sx__1kgv65vb Button_ButtonContainer_size_medium__1kgv65v9 Button_ButtonContainer_theme_dark__1kgv65v3 Button_ButtonContainer_variant_primary__1kgv65v5 Button_ButtonContainer_compound_0__1kgv65vd"
  data-rac=""
  type="button"
>
  <div
    class="Spinner_Container__gbijxg1 Spinner_Container_size_medium__gbijxg3"
  >
    <div
      class="Spinner_Circle__gbijxg5 Spinner_Circle_size_medium__gbijxgc Spinner_Circle_theme_dark__gbijxg6"
    />
    <div
      class="Spinner_Spin__gbijxgi Spinner_Spin_size_medium__gbijxgp Spinner_Spin_theme_dark__gbijxgm"
    />
  </div>
</button>
`;

exports[`Button component > renders with a size 1`] = `
<button
  class="Button_ButtonContainer__1kgv65v2 Button_ButtonContainer_iconLayout_sx__1kgv65vb Button_ButtonContainer_size_large__1kgv65va Button_ButtonContainer_theme_dark__1kgv65v3 Button_ButtonContainer_variant_primary__1kgv65v5 Button_ButtonContainer_compound_0__1kgv65vd"
  data-rac=""
  type="button"
>
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_l__3igy4nd Typography_BodyStyle_strong_true__3igy4ng"
  >
    Click me
  </p>
</button>
`;

exports[`Button component > renders with a theme 1`] = `
<button
  class="Button_ButtonContainer__1kgv65v2 Button_ButtonContainer_iconLayout_sx__1kgv65vb Button_ButtonContainer_size_medium__1kgv65v9 Button_ButtonContainer_theme_dark__1kgv65v3 Button_ButtonContainer_variant_primary__1kgv65v5 Button_ButtonContainer_compound_0__1kgv65vd"
  data-rac=""
  type="button"
>
  <p
    class="Typography__3igy4nb Typography_base__3igy4na Typography_BodyStyle_noMargin_true__3igy4nc Typography_BodyStyle_size_m__3igy4ne Typography_BodyStyle_strong_true__3igy4ng"
  >
    Click me
  </p>
</button>
`;
